metaflow/plugins/docker/docker_environment.py

```diff
@@ -0,0 +1,154 @@
+from metaflow.plugins.kubernetes.kubernetes_decorator import KubernetesDecorator
+from metaflow.plugins.pypi.conda_decorator import CondaStepDecorator
+from metaflow.plugins.pypi.pypi_decorator import PyPIStepDecorator
+
+
+class DockerEnvironmentException(MetaflowException):
+    headline = "Ran into an error while setting up the environment"
+
+    def __init__(self, msg):
+        super(DockerEnvironmentException, self).__init__(msg)
+
+
+class DockerEnvironment(MetaflowEnvironment):
+    TYPE = "docker"
+    _filecache = None
+
+    def __init__(self, flow):
+        self.steps_to_delegate = set()
+        self.flow = flow
+
+    def set_local_root(self, local_root):
+        # TODO: Make life simple by passing echo to the constructor and getting rid of
+        # this method's invocation in the decorator
+        self.local_root = local_root
+
+    def decospecs(self):
+        # Apply conda decorator to manage the task execution lifecycle.
+        return ("conda",) + super().decospecs()
+
+    def validate_environment(self, echo, datastore_type):
+        self.datastore_type = datastore_type
+        self.echo = echo
+
+        # Avoiding circular imports.
+        from metaflow.plugins import DATASTORES
+
+        self.datastore = [d for d in DATASTORES if d.TYPE == self.datastore_type][0]
+
+        # Use remote fast bakery for conda environments if configured.
+        if not _USE_BAKERY:
+            raise DockerEnvironmentException("Fast Bakery is not configured.")
+
+    def _init_conda_fallback(self):
+        # TODO: In the future we want to support executing with Docker even locally.
+        for step in self.flow:
+            if not any(_is_remote_deco(deco) for deco in step.decorators):
+                # We need to fall back to the Conda environmment for flows that are trying to execute anything locally.
+                self.steps_to_delegate.add(step.name)
+        if not self.steps_to_delegate:
+            return
+        # TODO: move to init if possible.
+        self.delegate = CondaEnvironment(self.flow)
+        self.delegate.set_local_root(self.local_root)
+
+    def init_environment(self, echo):
+        self._init_conda_fallback()
+        # First resolve environments through Conda, before PyPI.
+        echo("Baking Docker images for environment(s) ...")
+        for step in self.flow:
+            self.bake_image_for_step(step)
+        echo("Environments are ready!")
+        if self.steps_to_delegate:
+            # TODO: add debug echo to output steps that required a conda environment.
+            # The delegated conda environment also need to validate and init.
+            # we pass a set of steps we want to init to restrict the conda environments created.
+            self.delegate.validate_environment(echo, self.datastore_type)
+            self.delegate.init_environment(echo, self.steps_to_delegate)
+
+    def bake_image_for_step(self, step):
```

Comment:
multiple steps might share the same image - maybe we can check if that is the case and only bake unique images?

Reply:
the images are baked in sequence so the consecutive steps with identical packages should be hitting the cache at this point, skipping the baking. previously the caching was happening inside the fast_bakery, but moved to the environment side now so it should be more clear.

Reply:
rather than baking in sequence, we can perhaps bake them in parallel - that should save precious seconds